### DISCUSSION TITLE
Adding more support for Arrow Schema conversion with Dicts -> pa.map_()

### DISCUSSION THREAD

**Discussion by everettVT** [lancedb/lancedb#1658] [General]
Hi all, 

I feel like this is a pretty common use case, but a metadata field in pydantic generally looks something like this: 

```
class MyModel(BaseModel):
    id: int
    metadata: Dict[str,Any]
```

 lancedb.pydantic -> LanceModel only supports conversions for:
```
def _py_type_to_arrow_type(py_type: Type[Any], field: FieldInfo) -> pa.DataType:
    """Convert a field with native Python type to Arrow data type.

    Raises
    ------
    TypeError
        If the type is not supported.
    """
    if py_type is int:
        return pa.int64()
    elif py_type is float:
        return pa.float64()
    elif py_type is str:
        return pa.utf8()
    elif py_type is bool:
        return pa.bool_()
    elif py_type is bytes:
        return pa.binary()
    elif py_type is date:
        return pa.date32()
    elif py_type is datetime:
        tz = get_extras(field, "tz")
        return pa.timestamp("us", tz=tz)
    elif getattr(py_type, "__origin__", None) in (list, tuple):
        child = py_type.__args__[0]
        return pa.list_(_py_type_to_arrow_type(child, field))
    raise TypeError(
        f"Converting Pydantic type to Arrow Type: unsupported type {py_type}."
    )
```

I'd like to suggest we broaden support for dictionaries of a couple different types (ie Dict[str,str], Dict[str,int], Dict[str, float], ...) 

The corresponding arrow type would be a pa.map_().

What kind of  testing and integration would supporting this require?

**Comment by wjones127** [comment]
I think we'd be open to this. I'd warn that `MapArray` tends to be an afterthought in many of the Arrow SDKs, so could turn up some bugs with PyArrow or other libraries while implementing.

I think the main thing you want to test is that you can easily create a model with it and it can roundtrip data easily. I'm not sure much is required for testing beyond that.
